File: dataencryption/src/main/java/com/mkyong/crypto/encryptor/CryptoUtils.java
package com.mkyong.crypto.encryptor;

import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.KeySpec;

import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Derived From mkyong.com:
 * https://mkyong.com/java/java-aes-encryption-and-decryption/
 * And:
 * https://github.com/mkyong/core-java
 *
 * MIT License: https://github.com/mkyong/core-java/blob/master/LICENSE
 */
public class CryptoUtils {

    public byte[] getRandomNonce(int numBytes) {
        byte[] nonce = new byte[numBytes];
        new SecureRandom().nextBytes(nonce);

        return nonce;
    }

    // AES 256 bits secret key derived from a password
    public SecretKey getAESKeyFromPassword(char[] password, byte[] salt) throws NoSuchAlgorithmException, InvalidKeySpecException {
        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");

        // iterationCount = 65536, keyLength = 256
        KeySpec spec = new PBEKeySpec(password, salt, 65536, 256);
        return new SecretKeySpec(factory.generateSecret(spec).getEncoded(), "AES");
    }
}

File: dataencryption/src/main/java/com/mkyong/crypto/encryptor/CryptoUtils.kt
package com.mkyong.crypto.encryptor

import java.security.SecureRandom
import java.security.spec.KeySpec
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import javax.inject.Inject

/**
 * Derived From mkyong.com:
 * https://mkyong.com/java/java-aes-encryption-and-decryption/
 * And:
 * https://github.com/mkyong/core-java
 *
 * MIT License: https://github.com/mkyong/core-java/blob/master/LICENSE
 */
class CryptoUtils @Inject constructor() {
    companion object {
        const val ENCRYPT_ALGO = "AES/GCM/NoPadding"

        const val TAG_LENGTH_BIT = 128 // must be one of {128, 120, 112, 104, 96}
        const val IV_LENGTH_BYTE = 12
        const val SALT_LENGTH_BYTE = 16
    }

    fun getRandomNonce(numBytes: Int): ByteArray {
        val nonce = ByteArray(numBytes)
        SecureRandom().nextBytes(nonce)
        return nonce
    }

    // AES 256 bits secret key derived from a password
    fun getAESKeyFromPassword(password: CharArray?, salt: ByteArray?): SecretKey {
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")

        // iterationCount = 65536, keyLength = 256
        val spec: KeySpec = PBEKeySpec(password, salt, 65536, 256)
        return SecretKeySpec(factory.generateSecret(spec).encoded, "AES")
    }
}